aten/src/ATen/native/AutogradComposite.cpp

```diff
@@ -91,18 +93,85 @@ bool _has_same_storage_numel(const at::Tensor& base, const at::Tensor& other) {
   return base.storage().sym_nbytes() / base.itemsize() == other.storage().sym_nbytes() / other.itemsize();
 }
 
-Tensor _lazy_clone(Tensor const& self) {
+Tensor _lazy_clone(Tensor const& self, std::optional<c10::Device> device_opt) {
   c10::StorageImpl* self_storage = self.storage().unsafeGetStorageImpl();
-  c10::intrusive_ptr<c10::StorageImpl> storage =
-    c10::impl::cow::lazy_clone_storage(*self_storage);
+  c10::intrusive_ptr<c10::StorageImpl> storage = nullptr;
+
+  if (device_opt.has_value()) {
+    c10::Device src_device = self.device();
+    c10::Device dst_device = device_opt.value();
+    c10::DeviceType src_device_type = src_device.type();
+    c10::DeviceType dst_device_type = dst_device.type();
+    TORCH_CHECK(
+      (src_device_type == dst_device_type)
+      || (src_device_type == c10::kCPU && dst_device_type == c10::kMPS)
+      || (src_device_type == c10::kMPS && dst_device_type == c10::kCPU),
+      "Lazy cloning is only supported in the following cases: ",
+      "cpu to mps, ",
+      "mps to cpu, ",
+      "or between the same device. Got ", src_device, " to ", dst_device
+    );
+
+    if (src_device_type == dst_device_type) {
+      c10::DeviceIndex src_index = at::empty({}, at::TensorOptions().device(src_device)).device().index();
+      c10::DeviceIndex dst_index = at::empty({}, at::TensorOptions().device(dst_device)).device().index();
+
+      // NOTE: This case might already work, but we do not have any tests for it
```

Comment:
IIRC, for cuda, it will only work if you have p2p access enabled. For non-cuda, not sure.
We should definitely keep this check